### PR TITLE
Add `validate` to `SingleTableMetadata`

### LINKED ISSUE #879: Add validate method to SingleTableMetadata
### Problem Description

_As a user, it would be useful to be able to validate whether my metadata is formatted correctly_

### Expected behavior

- Add `validate` method
- Validation consists of validating three separate parts of the metadata. The full details are in the `Additional context` section.
    1. Validating the columns
    2. Validating the keys
    3. Validating the constraints
- If the metadata is not valid: Raises an `InvalidMetadataError` with a description of all the errors found.
```bash
>>> metadata.validate()
InvalidMetadataError: The metadata is not valid

Error: Invalid values ("pii") for datetime column "start_date".
Error: Invalid regex format string "[A-{6}" for text column "user_id"
Error: Unknown key value 'uuid'. Keys should be columns that exist in the table.
Error: A Unique constraint is being applied to column "user_id". This column is already a key for that table.
Error: Invalid increment value (0.5) in a FixedIncrements constraint. Increments must be positive integers.
```

### Additional context

- **Column validation**: Each `sdtype` has different validation rules. They are listed below
    - `numerical`
        - Required attributes are: `representation`
        - Throw an error if any attributes besides `representation` are present
        `Error: Invalid values ("pii") for numerical column "age".`
    - `datetime`
        - Required attributes are: `datetime_format`
        - "datetime_format" must be a valid, parsable format string
        `Error: Invalid datetime format string "%O" for datetime column "start_date"`
        - There should no other attributes present
        `Error: Invalid values ("pii") for datetime column "start_date".`
    - `categorical`
        - Required attributes are either: `order` or `order_by`
        - "order" and "order_by" cannot both be present. You can only have 0 or 1 of these attributes
        `Error: Categorical column "education" has both an "order" and "order_by" attribute. Only 1 is allowed.`
        - If present, "order_by" must be either set to "numerical_value" or "alphabetical"
        `Error: Unknown ordering method "testing" provided for categorical column "education". Ordering method must be "numerical_value" or "alphabetical"`
        - If present, "order" must be a list with 1 or more elements
        `Error: Invalid order value provided for categorical column "education". The "order" must be a list with 1 or more elements.`
        - No other attributes can be present
        `Error: Invalid values ("pii") for categorical column "education".`
    - `boolean`
        - No required attributes
        - Throw an error if any attributes are present
        `Error: Invalid value ("pii") for boolean column "is_subscribed".`
    - `text`
        - Required attributes are: `regex_format`
        - "regex_format" is present but the string isn't a valid regex string that can be parsed
        `Error: Invalid regex format string "[A-{6}" for text column "user_id"`
        - Throw an error if any other attributes are present
        `Error: Invalid values ("pii") for text column "user_id".`
    - `Real World (Semantic) Types` (ie. `phone_number`)
        - No required parameters
        - `pii` is an optional parameter
        - If "pii" exists but it is not True or False, throw an error
        `Error: Invalid pii value provided for phone_number column "user_cell". The "pii" value must be set to True or False.`
        - Throw an error if any other attributes are present
        `Error: Invalid values ("datetime_format") for phone_number column "user_cell".`
        - raise a warning if the `sdtype` isn't fully supported. 
        `Warning: sdtype 'location' is not fully supported. The SDV will model this as a categorical variable.`
        `Warning: sdtype 'location' is not fully supported. The SDV will anonymize this column using random characters.`

- **Key validation**
    - "primary_key" must be a string or list of strings
    - "sequence_key" must a string or list of strings
    - "alternate_keys" must be a list of strings or a nested list of strings
    - "sequence_index" must be a string
    - The strings must correspond to the column names as specified in the other part of the Metadata
    `Error: Unknown key value 'uuid'. Keys should be columns that exist in the table.`
    - "sequence_index" cannot be the same as "sequence_key"
    `Error: sequence_index and sequence_key have the same value ('patient_id'). These columns must be different.`

- **Constraint validation**
    - Use each constraints `_validate_inputs` method and surface those errors #878 
    - Check for the following errors still
    - `Unique`
        - Each of the column names in "column_names" must be a column that is present in the "columns" specification
        `Error: A Unique constraint is being applied to invalid column names ("age", "weight"). The columns must exist in the table.`
        - "column_names" must include at least 1 column that is NOT a primary key or alternate key. Primary keys and alternate keys will already be guaranteed to be unique, so there's no need to add it in as a constraint.
        `Error: A Unique constraint is being applied to column "age". This column is already a key for that table.`
    - `FixedCombinations`
        - Each of the column names in "colum_names" must be a column that is present in the "columns" specification
        `Error: A FixedCombinations constraint is being applied to invalid column names ("C", "D"). The columns must exist in the table.`
    - `Inequality`
        - The string in "high_column_name" and "low_column_name" must be a column that is present in the "columns" specification
        `Error: An Inequality constraint is being applied to invalid column names ("C", "D"). The columns must exist in the table.`
        - Both high and low columns must be either type "numerical" or type "datetime"
        `Error: An Inequality constraint is being applied to mismatched sdtypes ("C", "D"). Both columns must be either numerical or datetime.`
    - `ScalarInequality`
        - "column_name" must refer to a column in the table
        `Error: A ScalarInequality constraint is being applied to invalid column names ("C"). The columns must exist in the table.`
        - The "value" must make sense based on the column type
            - If the column is "numerical", then "value" must be an int or float
            - If the column is "datetime", then "value" must be a datetime string of the right format
            - No other types are compatible
	    `Error: A ScalarInequality constraint is being applied to mismatched sdtypes. Numerical columns must be compared to integer or float values. Datetimes column must be compared to datetime strings.`
    - `Range`
        - The strings in each of the column names must be a column that is present in the "columns" specification
        `Error: A Range constraint is being applied to invalid column names ("C", "D"). The columns must exist in the table.`
        - All columns must be either type "numerical" or type "datetime"
        `Error: A Range constraint is being applied to mismatched sdtypes ("C", "D", "E"). All columns must be either numerical or datetime.`
    - `ScalarRange`
        - "column_name" must refer to a column in the table
        `Error: A ScalarRange constraint is being applied to invalid column names ("C"). The columns must exist in the table.`
        - The high and low values must make sense based on the column type
            - If the column is "numerical", then the values must be floats/ints
            - If the column is "datetime", then the values must be a datetime string of the right format
            - No other types are compatible
        `Error: A ScalarRange constraint is being applied to mismatched sdtypes. Numerical columns must be compared to integer or float values. Datetimes column must be compared to datetime strings.`
    - `Positive`
        - "column_name" must refer to a column in the table
        `Error: A Positive constraint is being applied to invalid column names ("C"). The columns must exist in the table.`
        - Column name must type "numerical"
        `Error: A Positive constraint is being applied to an invalid column ("C"). This constraint is only defined for numerical columns.`
    - `Negative`
        - "column_name" must refer to a column in the table
        `Error: A Negative constraint is being applied to invalid column names ("C"). The columns must exist in the table.`
        - Column name must type "numerical"
        `Error: A Negative constraint is being applied to an invalid column ("C"). This constraint is only defined for numerical columns.`
    - `FixedIncrements`
        - Column name should refer to a column defined in the metadata
        `Error: A FixedIncrements constraint is being applied to invalid column names ("C"). The columns must exist in the table.`
    - OneHotEncoding
        - Column names must be valid columns (present in the "columns" part of the metadata)
        `Error: A OneHotEncoding constraint is being applied to invalid column names ("C", "D", "E"). The columns must exist in the table.`
    - `CustomConstraint`
        - Column names must be valid columns (present in the "columns part of the metadata)
        `Error: A <module>.<name> constraint is being applied to invalid column names ("C", "D"). The columns must exist in the table.`
    - Misc
        - If the constraint isn't found, throw an error:
        `Error: Invalid constraints ('Other'). `

### PR DESCRIPTION
**Pull request by fealho**
Resolve #879.

### CODE REVIEW

File: sdv/metadata/errors.py [hunk 1]
```diff
@@ -5,5 +5,9 @@ class MetadataError(Exception):
     """Error to raise when Metadata is not valid."""
 
 
+class InvalidMetadataError(Exception):
```
Comment (fealho) [inline sdv/metadata/errors.py:8]: Why create this new error? Couldn't we just use the `MetadataError` defined above?

File: sdv/metadata/single_table.py [hunk 2]
```diff
@@ -52,13 +52,16 @@ def _validate_numerical(self, column_name, **kwargs):
         representation = kwargs.get('representation')
         if representation and representation not in self._NUMERICAL_REPRESENTATIONS:
             raise ValueError(
-                f"Invalid value for 'representation' {representation} for column '{column_name}'.")
+                f"Invalid value for 'representation' '{representation}'"
```
Comment (fealho) [inline sdv/metadata/single_table.py:55]: It was missing the quotes around `{representation}`

File: sdv/metadata/single_table.py [hunk 3]
```diff
@@ -52,13 +52,16 @@ def _validate_numerical(self, column_name, **kwargs):
...
+                f" for column '{column_name}'."
+            )
 
     @staticmethod
     def _validate_datetime(column_name, **kwargs):
         datetime_format = kwargs.get('datetime_format')
-        if datetime_format:
+        if datetime_format is not None:
```
Comment (fealho) [inline sdv/metadata/single_table.py:62]: It was not validating if the value was an empty string.

File: sdv/metadata/single_table.py [hunk 4]
```diff
@@ -52,13 +52,16 @@ def _validate_numerical(self, column_name, **kwargs):
...
             try:
+                # NOTE: I don't know if this ever crashes, it just returns the string as is
```
Comment (fealho) [inline sdv/metadata/single_table.py:64]: @pvk-developer I'm not sure how `strftime` works, but it doesn't seem to raise errors even if you pass it a normal string like 'value'

File: tests/unit/metadata/test_single_table.py [hunk 5]
```diff
@@ -877,27 +877,53 @@ def test_set_primary_key_validation(self):
...
+
+        Side Effect:
+            - A ``ValueError`` should be raised.
+        """
+        # Setup
+        instance = SingleTableMetadata()
+        instance._columns = {'a', 'd'}
+
+        err_msg = (
+            "Unknown primary key values {'b', 'c'}."
```
Comment (fealho) [inline tests/unit/metadata/test_single_table.py:897]: @pvk-developer @amontanez24 How did we decide to tests that sets match?

**Comment by fealho** [comment]
The #879 issue asks us to test `Real World (Semantic) Types (ie. phone_number`, but this was not done in the `validate_columns` PR. Do we still want to add this?

File: sdv/constraints/base.py [hunk 6]
```diff
@@ -126,7 +126,8 @@ def _validate_inputs(cls, **kwargs):
                 f'Invalid values {invalid_vals} are present in {article} {constraint} constraint.'
             ))
 
-        raise MultipleConstraintsErrors(errors)
+        if errors:
```
Comment (fealho) [inline sdv/constraints/base.py:129]: Unrelated.

File: sdv/constraints/base.py [hunk 7]
```diff
@@ -135,16 +136,16 @@ def _validate_metadata_columns(cls, metadata, **kwargs):
         else:
             column_names = kwargs.get('column_names')
 
-        missing_columns = set(column_names) - set(metadata._columns)
-
+        missing_columns = set(column_names) - set(metadata._columns) - {None}
```
Comment (fealho) [inline sdv/constraints/base.py:139]: Unrelated. This happens when a column name was not passed/is missing.

File: sdv/constraints/base.py [hunk 8]
```diff
@@ -135,16 +136,16 @@ def _validate_metadata_columns(cls, metadata, **kwargs):
...
         if missing_columns:
+            article = 'An' if cls.__name__ == 'Inequality' else 'A'
             raise ConstraintMetadataError(
-                f'A {cls.__name__} constraint is being applied to invalid column names '
+                f'{article} {cls.__name__} constraint is being applied to invalid column names '
```
Comment (fealho) [inline sdv/constraints/base.py:143]: Unrelated.

File: sdv/constraints/base.py [hunk 9]
```diff
@@ -135,16 +136,16 @@ def _validate_metadata_columns(cls, metadata, **kwargs):
...
                 f'{missing_columns}. The columns must exist in the table.'
             )
 
-    @classmethod
-    def _validate_metadata_specific_to_constraint(cls, metadata, **kwargs):
+    @staticmethod
```
Comment (fealho) [inline sdv/constraints/base.py:147]: Unrelated. I'm not sure why we were using classmethod? It seems to me that all child classes can be set to staticmethods, so I did so.

File: sdv/metadata/single_table.py [hunk 10]
```diff
@@ -250,55 +248,90 @@ def _validate_datatype(id):
         """Check whether id is a string or a tuple of strings."""
         return isinstance(id, str) or isinstance(id, tuple) and all(isinstance(i, str) for i in id)
 
+    def _validate_key(self, id, key_type):
```
Comment (fealho) [inline sdv/metadata/single_table.py:251]: The previous implementation had the validation inside the set_key methods. Pulling it out of the key methods so I can use it in the `validate` method.

File: sdv/metadata/single_table.py [hunk 11]
```diff
@@ -307,15 +340,81 @@ def set_sequence_index(self, column_name):
...
+        sequence_key = {sk} if isinstance(sk, str) else set(sk)
+        if self._sequence_index in sequence_key or sk is None:
+            index = {self._sequence_index}
+            raise ValueError(
+                f"'sequence_index' and 'sequence_key' have the same value {index}."
+                ' These columns must be different.'
+            )
 
-        self._metadata['sequence_index'] = column_name
+    def _validate_constraint(self, constraint_name, **kwargs):
```
Comment (fealho) [inline sdv/metadata/single_table.py:357]: Pulling this out of `add_constraint` because I need it in for `validate`.

File: sdv/metadata/single_table.py [hunk 12]
```diff
@@ -307,15 +340,81 @@ def set_sequence_index(self, column_name):
...
+            - ``InvalidMetadataError`` if the metadata is invalid.
+        """
+        # Validate constraints
+        errors = []
+        for tuple in self._constraints:
+            constraint_name, kwargs = tuple
+            try:
+                self._validate_constraint(constraint_name, **kwargs)
+            except MultipleConstraintsErrors as e:
+                reformated_errors = '\n'.join(map(str, e.errors))
```
Comment (fealho) [inline sdv/metadata/single_table.py:393]: Need to do this because `MultipleConstraintsErrors` sets it's own `__str__`, which messes up the formatting of the error message.

File: sdv/metadata/single_table.py [hunk 13]
```diff
@@ -307,15 +340,81 @@ def set_sequence_index(self, column_name):
...
+            raise InvalidMetadataError(
+                'The following errors were found in the metadata:\n\n'
+                + '\n'.join([str(e) for e in errors])
+            )
 
     def to_dict(self):
         """Return a python ``dict`` representation of the ``SingleTableMetadata``."""
         metadata = {}
-        for key, value in self._metadata.items():
+        for key in self._KEYS:
```
Comment (fealho) [inline sdv/metadata/single_table.py:416]: Change needed because we deleted _metadata.

File: tests/unit/constraints/test_tabular.py [hunk 14]
```diff
@@ -990,7 +990,7 @@ def test__validate_metadata_specific_to_constraint_datetime_error(self):
 
         # Run
         error_message = re.escape(
-            'An Inequality constraint is being applied to mismatched sdtypes '
+            'An Inequality constraint is being applied to mismatched sdtype columns '
```
Comment (fealho) [inline tests/unit/constraints/test_tabular.py:993]: Unrelated. Phrasing was confusing, it made it seem the values were sdtypes when they are actually columns.

**Review by fealho — COMMENTED** [review]
Left a few comments to help understand the code. Messages marked with `Unrelated` mean they are standalone and unrelated to the PR itself.

File: sdv/constraints/base.py (hunk 6 above)
Comment (amontanez24) [reply sdv/constraints/base.py:129]: Good catch! maybe we should add a test for this case

File: sdv/constraints/base.py (hunk 9 above)
Comment (amontanez24) [reply sdv/constraints/base.py:147]: should we change it to match in the base as well?

File: tests/unit/constraints/test_tabular.py (hunk 14 above)
Comment (amontanez24) [reply tests/unit/constraints/test_tabular.py:993]: minor: can we reword to `is being applied to columns with mismatched sdtypes` then? I think that makes more sense

File: tests/unit/constraints/test_tabular.py [hunk 15]
```diff
@@ -1035,7 +1035,7 @@ def test__validate_metadata_specific_to_constraint_numerical_error(self):
...
```
Comment (amontanez24) [inline tests/unit/constraints/test_tabular.py:1038]: same

File: tests/unit/constraints/test_tabular.py [hunk 16]
```diff
@@ -2581,7 +2581,7 @@ def test__validate_metadata_specific_to_constraint_datetime_error(self):
 
         # Run
         error_message = re.escape(
-            'A Range constraint is being applied to mismatched sdtypes '
+            'A Range constraint is being applied to mismatched sdtype columns '
```
Comment (amontanez24) [inline tests/unit/constraints/test_tabular.py:2584]: same

File: tests/unit/constraints/test_tabular.py [hunk 17]
```diff
@@ -2637,7 +2637,7 @@ def test__validate_metadata_specific_to_constraint_numerical_error(self):
...
```
Comment (amontanez24) [inline tests/unit/constraints/test_tabular.py:2640]: same

File: tests/unit/metadata/test_single_table.py [hunk 18]
```diff
@@ -914,89 +938,89 @@ def test_set_primary_key_warning(self, warning_mock):
...
-        err_msg = "'sequence_key' must be a string or tuple of strings."
+        err_msg = (
+            "Unknown sequence key values {'b'}."
+            ' Keys should be columns that exist in the table.'
+        )
         # Run / Assert
         with pytest.raises(ValueError, match=err_msg):
-            instance.set_sequence_key(('1', 2, '3'))
+            instance.set_sequence_key(('a', 'b', 'd'))
+            # NOTE: used to be ('a', 'b', 'd', 'c')
```
Comment (amontanez24) [inline tests/unit/metadata/test_single_table.py:992]: what is this?

File: tests/unit/metadata/test_single_table.py [hunk 19]
```diff
@@ -1033,16 +1112,100 @@ def test_set_sequence_index_validation(self):
...
+
+        err_msg = (
+            "'sequence_index' and 'sequence_key' have the same value {'abc'}."
+            ' These columns must be different.'
+        )
+        # Run / Assert
+        with pytest.raises(ValueError, match=err_msg):
+            instance._validate_sequence_index_not_in_sequence_key()
+
+    def test_validate(self):
```
Comment (amontanez24) [inline tests/unit/metadata/test_single_table.py:1166]: I think we should have some of the mocks raise errors and assure that the error messages are correctly concatenated

File: tests/unit/metadata/test_single_table.py [hunk 20]
```diff
@@ -872,39 +867,68 @@ def test_set_primary_key_validation(self):
...
+        instance._columns = {'a', 'd'}
+
+        err_msg = (
+            "Unknown primary key values {'b'}."
+            ' Keys should be columns that exist in the table.'
+        )
+        # Run / Assert
+        with pytest.raises(ValueError, match=err_msg):
+            instance.set_primary_key(('a', 'b', 'd'))
+            # NOTE: used to be ('a', 'b', 'd', 'c')
```
Comment (amontanez24) [inline tests/unit/metadata/test_single_table.py:900]: what are these notes for?

**Review by amontanez24 — CHANGES_REQUESTED** [review]
Looking good. A few minor comments

File: tests/unit/metadata/test_single_table.py (hunk 20 above)
Comment (fealho) [reply tests/unit/metadata/test_single_table.py:900]: For Plamen to know where it used to be a set of errors, to change back after we decide how to handle them.

File: tests/unit/metadata/test_single_table.py (hunk 19 above)
Comment (fealho) [reply tests/unit/metadata/test_single_table.py:1166]: That's why I created the integration test `test_validate_errors`, which prints the big error message with basically all the errors, and checks that everything is formatted properly. Any problem with the errors will more easily be caught there than through any mocks I feel.

Let me know if you still think this change necessary.

**Comment by fealho** [comment]
@amontanez24 I'm not sure what you mean here: https://github.com/sdv-dev/SDV/pull/930#discussion_r936945748, it was already changed in the base.

File: tests/unit/metadata/test_single_table.py (hunk 19 above)
Comment (amontanez24) [reply tests/unit/metadata/test_single_table.py:1166]: It doesn't need to be as in depth, but for test coverage purposes I think we need it. The following line only gets entered if there are errors
https://github.com/sdv-dev/SDV/pull/930/files#diff-c15919d0f7ddead675009beb9282fab906be3b8b3232e74ec347332563dcf4faR407

File: tests/unit/constraints/test_base.py [hunk 21]
```diff
@@ -159,6 +159,21 @@ def test_import_object_function():
...
+        The method should only raise errors if the input paramaters are invalid.
+
+        Raise:
+            - ``MultipleConstraintsErrors`` if errors were found.
+        """
+        # Run
+        Constraint._validate_inputs(args='value', kwargs='value')
+
+        # Run / Assert
+        with pytest.raises(MultipleConstraintsErrors):
```
Comment (amontanez24) [inline tests/unit/constraints/test_base.py:174]: minor: maybe we should match the error like we've been doing to avoid more changes when we add linting

File: sdv/metadata/single_table.py [hunk 22]
```diff
@@ -307,15 +340,81 @@ def set_sequence_index(self, column_name):
...
         """
-        if not isinstance(column_name, str):
-            raise ValueError("'sequence_index' must be a string.")
+        self._validate_sequence_index(column_name)
+        self._sequence_index = column_name
+
+    def _validate_sequence_index_not_in_sequence_key(self):
+        """Check that ``_sequence_index`` and ``_sequence_key`` don't overlap."""
+        sk = self._sequence_key
+        sequence_key = {sk} if isinstance(sk, str) else set(sk)
```
Comment (amontanez24) [inline sdv/metadata/single_table.py:349]: minor: I think we've been trying to avoid letter abbreviations for variables. If `sequence_key` fits I would say use that instead of sk. Idk if the lint will complain about that later as well

File: sdv/metadata/single_table.py [hunk 23]
```diff
@@ -307,15 +340,81 @@ def set_sequence_index(self, column_name):
...
+        self._append_error(errors, self._validate_alternate_keys, self._alternate_keys)
+        self._append_error(errors, self._validate_sequence_index, self._sequence_index)
+        self._append_error(errors, self._validate_sequence_index_not_in_sequence_key)
+
+        # Validate columns
+        for column, kwargs in self._columns.items():
+            self._append_error(errors, self._validate_column, column, **kwargs)
+
+        if errors:
+            raise InvalidMetadataError(
```
Comment (amontanez24) [inline sdv/metadata/single_table.py:408]: Not necessary for this PR but later I think we might want to make a custom error that takes a list of errors and joins them in the message. We're probably going to be doing that a lot in the future

**Review by amontanez24 — APPROVED** [review]
Some very minor comments, but I think this is good to go!

**Review by pvk-developer — APPROVED** [review]
